core/templates/components/state-editor/state-responses-editor/state-responses.component.spec.ts

```diff
@@ -0,0 +1,996 @@
+      ' or direct the learner to a different card.');
+  });
+
+  it('should open add response modal when user clicks on' +
+    ' \'+ ADD RESPONSE\' button', () => {
+    spyOn($uibModal, 'open').and.callThrough();
+
+    $scope.openAddAnswerGroupModal();
+
+    expect($uibModal.open).toHaveBeenCalled();
+  });
+
+  it('should open add response modal and save new answer groups' +
+    ' added by the user', () => {
+    spyOn(ResponsesService, 'save').and.callFake(
+      (answerGroups, defaultOutcome, callback) => {
+        callback(answerGroups, defaultOutcome);
+      }
+    );
+    // Returning rejecting callback as the modal opens again, as reopen is true
+    // so we close it when it is opened for the second time.
+    spyOn($uibModal, 'open').and.returnValues({
+      result: $q.resolve({
+        reopen: true,
+        tmpRule: new Rule('', null, null),
+        tmpOutcome: outcomeObjectFactory
+          .createNew('Hola', '1', 'Feedback text', []),
+        tmpTaggedSkillMisconceptionId: ''
+      })
+    }, {
+      result: $q.reject()
+    });
+    $scope.answerGroups = [];
+
+    $scope.openAddAnswerGroupModal();
+    $scope.$apply();
+
+    expect($scope.answerGroups).toEqual([answerGroupObjectFactory.createNew(
+      [new Rule('', null, null)], outcomeObjectFactory.createNew(
+        'Hola', '1', 'Feedback text', []), [], ''
+    )]);
+    expect(ResponsesService.save).toHaveBeenCalled();
+  });
+
+  it('should clear warnings when modal is closed', () => {
+    spyOn(AlertsService, 'clearWarnings');
+    spyOn($uibModal, 'open').and.returnValue({
+      result: $q.reject()
+    });
+
+    $scope.openAddAnswerGroupModal();
+    $scope.$apply();
+
+    expect(AlertsService.clearWarnings).toHaveBeenCalledTimes(2);
+  });
+
+  it('should open delete answer group modal when user clicks' +
+    ' on delete button', () => {
+    spyOn($uibModal, 'open').and.callThrough();
+
+    $scope.deleteAnswerGroup(0, new Event(''));
+
+    expect($uibModal.open).toHaveBeenCalled();
+  });
+
+  it('should delete answer group after modal is opened', () => {
+    spyOn(ResponsesService, 'deleteAnswerGroup').and.callFake(
+      (index, callback) => {
+        callback();
+      }
+    );
+    spyOn($uibModal, 'open').and.returnValue({
+      result: $q.resolve()
+    });
+
+    $scope.deleteAnswerGroup(0, new Event(''));
+    $scope.$apply();
+
+    expect($uibModal.open).toHaveBeenCalled();
+    expect(ResponsesService.deleteAnswerGroup).toHaveBeenCalled();
+  });
+
+  it('should clear warnings when delete answer group modal is closed', () => {
+    spyOn(AlertsService, 'clearWarnings');
+    spyOn($uibModal, 'open').and.returnValue({
+      result: $q.reject()
+    });
+
+    $scope.deleteAnswerGroup(0, new Event(''));
+    $scope.$apply();
+
+    expect(AlertsService.clearWarnings).toHaveBeenCalledTimes(2);
+  });
+
+  it('should update active answer group for newly tagged misconception', () => {
```

Comment:
Please separate set-up, actions, and post-checks with a line in between. Please do this for the rest of the tests

Reply:
Done